File: sdks/python/src/untrace/exporter.py
"""Exporter implementation for the Untrace SDK."""

import asyncio
import json
import logging
from typing import Any, Dict, List, Optional, Union
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult, ReadableSpan
from opentelemetry.trace import SpanKind, StatusCode

from .types import UntraceConfig


logger = logging.getLogger(__name__)


class UntraceExporter(SpanExporter):
    """Exporter that sends spans to Untrace backend."""

    def __init__(self, config: UntraceConfig):
        """Initialize the Untrace exporter.

        Args:
            config: Untrace configuration
        """
        self.config = config
        self._export_url = f"{config.base_url}/api/v1/traces/ingest"

        if config.debug:
            logger.info(
                "[UntraceExporter] Initialized with config: %s",
                {
                    "apiKey": f"{config.api_key[:8]}...{config.api_key[-4:]}" if config.api_key else "missing",
                    "baseUrl": config.base_url,
                    "debug": config.debug,
                }
            )

    def export(self, spans: List[ReadableSpan]) -> SpanExportResult:
        """Export spans to Untrace backend.

        Args:
            spans: List of spans to export

        Returns:
            Export result
        """
        if not spans:
            return SpanExportResult.SUCCESS

        try:
            # Run the async export in a new event loop
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            try:
                result = loop.run_until_complete(self._export_spans(spans))
                return result
            finally:
                loop.close()
        except Exception as error:
            logger.error("[UntraceExporter] Export failed: %s", error)
            return SpanExportResult.FAILURE

    async def _export_spans(self, spans: List[ReadableSpan]) -> SpanExportResult:
        """Async export implementation.

        Args:
            spans: List of spans to export

        Returns:
            Export result
        """
        try:
            if self.config.debug:
                logger.info(
                    "[UntraceExporter] Exporting %d spans to %s",
                    len(spans),
                    self._export_url
                )
                logger.info(
                    "[UntraceExporter] Span names: %s",
                    [span.name for span in spans]
                )

            payload = self._convert_spans_to_payload(spans)

            if self.config.debug:
                logger.info("[UntraceExporter] Payload prepared, sending request...")

            import httpx

            async with httpx.AsyncClient() as client:
                response = await client.post(
                    self._export_url,
                    json=payload,
                    headers={
                        "Authorization": f"Bearer {self.config.api_key}",
                        "Content-Type": "application/json",
                        **self.config.headers,
                    },
                    timeout=30.0,
                )

                if not response.is_success:
                    logger.error(
                        "[UntraceExporter] Export failed with status %d: %s",
                        response.status_code,
                        response.text
                    )
                    return SpanExportResult.FAILURE

                if self.config.debug:
                    logger.info("[UntraceExporter] Export successful")
                    logger.info("[UntraceExporter] Response: %s", response.text)

                return SpanExportResult.SUCCESS

        except Exception as error:
            logger.error("[UntraceExporter] Export failed: %s", error)
            return SpanExportResult.FAILURE

    def shutdown(self) -> None:
        """Shutdown the exporter."""
        # No cleanup needed for this exporter
        pass

    def force_flush(self, timeout_millis: Optional[int] = None) -> bool:
        """Force flush any pending spans.

        Args:
            timeout_millis: Timeout in milliseconds

        Returns:
            True if flush was successful
        """
        # No buffering, so nothing to flush
        return True

    def _convert_spans_to_payload(self, spans: List[ReadableSpan]) -> Dict[str, Any]:
        """Convert spans to OTLP JSON format.

        Args:
            spans: List of spans to convert

        Returns:
            OTLP payload dictionary
        """
        # Group spans by resource and scope
        resource_map: Dict[str, Dict[str, List[ReadableSpan]]] = {}

        for span in spans:
            # Convert attributes to regular dict for JSON serialization
            resource_attrs = dict(span.resource.attributes) if hasattr(span.resource.attributes, 'items') else span.resource.attributes
            resource_key = json.dumps(resource_attrs, sort_keys=True)
            scope_key = f"{span.instrumentation_scope.name}:{span.instrumentation_scope.version or ''}"

            if resource_key not in resource_map:
                resource_map[resource_key] = {}

            if scope_key not in resource_map[resource_key]:
                resource_map[resource_key][scope_key] = []

            resource_map[resource_key][scope_key].append(span)

        # Convert to OTLP format
        resource_spans = []

        for resource_key, scope_map in resource_map.items():
            # Get the first span to extract resource info
            first_span = next(iter(next(iter(scope_map.values()))))

            resource = {
                "attributes": self._convert_attributes(first_span.resource.attributes)
            }

            scope_spans = []

            for scope_key, span_list in scope_map.items():
                scope_name, scope_version = scope_key.split(":", 1)

                scope = {
                    "name": scope_name,
                    "version": scope_version if scope_version else None
                }

                scope_spans.append({
                    "scope": scope,
                    "spans": [self._convert_span(span) for span in span_list]
                })

            resource_spans.append({
                "resource": resource,
                "scopeSpans": scope_spans
            })

        return {"resourceSpans": resource_spans}

    def _convert_span(self, span: ReadableSpan) -> Dict[str, Any]:
        """Convert a single span to OTLP format.

        Args:
            span: Span to convert

        Returns:
            OTLP span dictionary
        """
        return {
            "traceId": format(span.context.trace_id, '032x'),
            "spanId": format(span.context.span_id, '016x'),
            "parentSpanId": format(span.parent.span_id, '016x') if span.parent else None,
            "name": span.name,
            "kind": span.kind.value,
            "startTimeUnixNano": str(self._time_to_nanos(span.start_time)),
            "endTimeUnixNano": str(self._time_to_nanos(span.end_time)),
            "attributes": self._convert_attributes(span.attributes),
            "status": {
                "code": span.status.status_code.value,
                "message": span.status.description
            },
            "events": [
                {
                    "timeUnixNano": str(self._time_to_nanos(event.timestamp)),
                    "name": event.name,
                    "attributes": self._convert_attributes(event.attributes or {})
                }
                for event in span.events
            ],
            "links": [
                {
                    "traceId": format(link.context.trace_id, '032x'),
                    "spanId": format(link.context.span_id, '016x'),
                    "attributes": self._convert_attributes(link.attributes or {})
                }
                for link in span.links
            ]
        }

    def _convert_attributes(self, attributes: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Convert attributes to OTLP format.

        Args:
            attributes: Attributes dictionary

        Returns:
            List of OTLP attributes
        """
        result = []

        # Convert BoundedAttributes to regular dict if needed
        if hasattr(attributes, 'items'):
            attrs_dict = dict(attributes)
        else:
            attrs_dict = attributes

        for key, value in attrs_dict.items():
            if value is None:
                continue

            attr = {"key": key, "value": self._convert_value(value)}
            result.append(attr)

        return result

    def _convert_value(self, value: Any) -> Dict[str, Any]:
        """Convert an attribute value to OTLP format.

        Args:
            value: Attribute value

        Returns:
            OTLP value dictionary
        """
        if isinstance(value, str):
            return {"stringValue": value}
        elif isinstance(value, bool):
            return {"boolValue": value}
        elif isinstance(value, int):
            return {"intValue": value}
        elif isinstance(value, float):
            return {"doubleValue": value}
        elif isinstance(value, (list, tuple)):
            return {
                "arrayValue": {
                    "values": [self._convert_value(v) for v in value]
                }
            }
        elif isinstance(value, dict):
            return {
                "kvlistValue": {
                    "values": [
                        {"key": k, "value": self._convert_value(v)}
                        for k, v in value.items()
                        if v is not None
                    ]
                }
            }
        else:
            # Fallback to string
            return {"stringValue": str(value)}

    def _time_to_nanos(self, timestamp: int) -> int:
        """Convert timestamp to nanoseconds.

        Args:
            timestamp: Timestamp in nanoseconds

        Returns:
            Timestamp in nanoseconds
        """
        return timestamp

File: sdks/go/untrace/exporter.go
package untrace

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
)

// UntraceExporter represents a custom exporter for Untrace
type UntraceExporter struct {
	config     Config
	httpClient *http.Client
	baseURL    string
}

// NewUntraceExporter creates a new Untrace exporter
func NewUntraceExporter(config Config) (*UntraceExporter, error) {
	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	return &UntraceExporter{
		config:     config,
		httpClient: client,
		baseURL:    config.BaseURL + "/v1/traces",
	}, nil
}

// ExportSpans exports spans to the Untrace API
func (e *UntraceExporter) ExportSpans(ctx context.Context, spans []sdktrace.ReadOnlySpan) error {
	if len(spans) == 0 {
		return nil
	}

	// Convert spans to the format expected by Untrace API
	payload, err := e.convertSpansToPayload(spans)
	if err != nil {
		return fmt.Errorf("failed to convert spans: %w", err)
	}

	// Send to Untrace API
	return e.sendToAPI(ctx, payload)
}

// Shutdown shuts down the exporter
func (e *UntraceExporter) Shutdown(ctx context.Context) error {
	// Nothing to shutdown for HTTP client
	return nil
}

// convertSpansToPayload converts OpenTelemetry spans to Untrace API format
func (e *UntraceExporter) convertSpansToPayload(spans []sdktrace.ReadOnlySpan) (map[string]interface{}, error) {
	// This is a simplified conversion - in a real implementation,
	// you would convert the spans to the exact format expected by Untrace API
	convertedSpans := make([]map[string]interface{}, 0, len(spans))

	for _, span := range spans {
		convertedSpan := map[string]interface{}{
			"trace_id":    span.SpanContext().TraceID().String(),
			"span_id":     span.SpanContext().SpanID().String(),
			"name":        span.Name(),
			"start_time":  span.StartTime().UnixNano(),
			"end_time":    span.EndTime().UnixNano(),
			"attributes":  span.Attributes(),
			"status":      span.Status(),
		}

		if span.Parent().SpanID().IsValid() {
			convertedSpan["parent_span_id"] = span.Parent().SpanID().String()
		}

		convertedSpans = append(convertedSpans, convertedSpan)
	}

	return map[string]interface{}{
		"spans": convertedSpans,
	}, nil
}

// sendToAPI sends the payload to the Untrace API
func (e *UntraceExporter) sendToAPI(ctx context.Context, payload map[string]interface{}) error {
	jsonData, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("failed to marshal payload: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", e.baseURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	// Set headers
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+e.config.APIKey)
	req.Header.Set("User-Agent", "untrace-sdk-go/0.1.0")

	// Add custom headers
	for key, value := range e.config.Headers {
		req.Header.Set(key, value)
	}

	resp, err := e.httpClient.Do(req)
	if err != nil {
		return &APIError{
			UntraceError: UntraceError{
				Message: "failed to send request to Untrace API",
				Err:     err,
			},
		}
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return NewAPIError(
			fmt.Sprintf("API request failed with status %d", resp.StatusCode),
			resp.StatusCode,
			string(body),
			nil,
		)
	}

	return nil
}

// CreateOTLPExporter creates an OTLP exporter configured for Untrace
func CreateOTLPExporter(config Config) (otlptrace.Client, error) {
	// Create HTTP client with custom headers
	client := otlptracehttp.NewClient(
		otlptracehttp.WithEndpoint(config.BaseURL),
		otlptracehttp.WithHeaders(map[string]string{
			"Authorization": "Bearer " + config.APIKey,
			"User-Agent":    "untrace-sdk-go/0.1.0",
		}),
	)

	return client, nil
}

// CreateResource creates an OpenTelemetry resource for Untrace
func CreateResource(config Config) *resource.Resource {
	attrs := []attribute.KeyValue{
		semconv.ServiceNameKey.String(config.ServiceName),
		semconv.ServiceVersionKey.String(config.Version),
		semconv.DeploymentEnvironmentKey.String(config.Environment),
	}

	// Add custom resource attributes
	for key, value := range config.ResourceAttributes {
		if str, ok := value.(string); ok {
			attrs = append(attrs, attribute.String(key, str))
		} else if num, ok := value.(int); ok {
			attrs = append(attrs, attribute.Int(key, num))
		} else if num, ok := value.(float64); ok {
			attrs = append(attrs, attribute.Float64(key, num))
		} else if b, ok := value.(bool); ok {
			attrs = append(attrs, attribute.Bool(key, b))
		}
	}

	return resource.NewWithAttributes(
		semconv.SchemaURL,
		attrs...,
	)
}
